### DISCUSSION TITLE
In emacs master: void-variable native-comp-eln-load-path

### DISCUSSION THREAD

**Discussion by meedstrom** [emacscollective/no-littering#245] [General]
Don't know if it's just me, but in emacs master as of Sep 7, the suggested early-init snippet (from readme https://github.com/emacscollective/no-littering)

```elisp
(when (fboundp 'startup-redirect-eln-cache)
  (startup-redirect-eln-cache
   (convert-standard-filename
    (expand-file-name  "var/eln-cache/" user-emacs-directory))))
```

causes 

```
Debugger entered--Lisp error: (void-variable native-comp-eln-load-path)
  startup-redirect-eln-cache("/home/kept/emacs/conf-alpaka/var/eln-cache/")
  (progn (startup-redirect-eln-cache (file-name-as-directory (file-name-concat user-emacs-directory "var" "eln-cache"))))
  (if (fboundp 'startup-redirect-eln-cache) (progn (startup-redirect-eln-cache (file-name-as-directory (file-name-concat user-emacs-directory "var" "eln-cache")))))
  eval-buffer(#<buffer  *load*> nil "/home/kept/emacs/conf-alpaka/early-init.el" nil t)  ; Reading at buffer position 1756
  load-with-code-conversion("/home/kept/emacs/conf-alpaka/early-init.el" "/home/kept/emacs/conf-alpaka/early-init.el" t t)
  load("/home/kept/emacs/conf-alpaka/early-init" noerror nomessage)
  #f(compiled-function () #<bytecode -0x13431c41b2974cbb>)()
  #f(compiled-function () #<bytecode -0x4307d7bb01d2857>)()
  handler-bind-1(#f(compiled-function () #<bytecode -0x4307d7bb01d2857>) (error) startup--debug)
  startup--load-user-init-file(#f(compiled-function () #<bytecode 0xca36a5e740c655d>))
  command-line()
  normal-top-level()
```

I compiled a fresh Git checkout of Emacs, on Debian sid without any custom configure flags, and used a fresh `--init-directory` that contains only aforementioned snippet.

**Comment by tarsius** [comment]
First check if `native-comp-available-p` returns non-nil.

`startup-redirect-eln-cache` tries to use `native-comp-eln-load-path` without ensuring it is actually defined. The code snippet should probably be extended to:

```elisp
(when (and (boundp 'native-comp-eln-load-path)
           (fboundp 'startup-redirect-eln-cache))
  (startup-redirect-eln-cache
   (convert-standard-filename
    (expand-file-name  "var/eln-cache/" user-emacs-directory))))
```

That addresses the issue when native-compilation is not actually available (by making this a no-op).

I currently have native compilation disabled (due to unrelated complications), so the rest is speculation. I would expect the original snippet to work as-is, provided native-compilation is available, but if it doesn't, then putting a `(require 'native-comppile nil t)` in front of it, *might* just do the trick.

**Comment by meedstrom** [comment]
Indeed `(native-comp-available-p)` returns nil. `(require 'native-compile nil t)`  too. That is a surprise. I'll guess the `./configure` script decided against building it for some reason.

Thanks.

But, not sure you need `(fboundp 'startup-redirect-eln-cache)`.

**Reply by meedstrom** [reply]
(Well, I guess yes for older emacsen)

**Reply by tarsius** [reply]
(Yes.)

I've updated the snippet.
